viewing_party/party.py

```diff
@@ -1,23 +1,184 @@
+    user_data["watched"].append(movie_dict)
+    return user_data
+
+
+def add_to_watchlist(user_data, movie_dict):
+    user_data["watchlist"].append(movie_dict)
+    return user_data
+
+
+
+def watch_movie(user_data, title):  
+    watch_list = user_data["watchlist"]
+    for movie in watch_list:
+        if (movie["title"]) == title:
+            watch_list.remove(movie)
+            user_data["watched"].append(movie)
+            return user_data
+    return user_data
+
+
+
 
 # -----------------------------------------
 # ------------- WAVE 2 --------------------
 # -----------------------------------------
+def get_watched_avg_rating(user_data):
+    rating_list = []
+    ratings_sum = 0
+    watch_list = user_data["watched"]
+    # access the list inside the dict
+    if len(watch_list) == 0:   # check if the list is empty
+        return 0.0
+    else: 
+        for i in range(len(watch_list)):    
+            rating_list.append(watch_list[i]["rating"])
+            # a list contains elements, each elements as a key-value pair
+            # watch_list[i]["rating"] only access the value from the key names "rating"
+    for rate in rating_list:
+        # access every rate inside the rating list
+        ratings_sum += rate
+        # add the rates together
+    average = ratings_sum/len(rating_list)
+    return average
+    
+    
+def get_most_watched_genre(user_data):
+    popular_genre = []
+    genre_list = user_data["watched"]
+    # print(genre_list)
+    if len(genre_list) == 0:
+        return None
+    else:
+       for j in range (len(genre_list)):
+            popular_genre.append(genre_list[j]["genre"])
+            # print(popular_genre)
+    return max(set(popular_genre), key = popular_genre.count)
+
+
 
 
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
 
+def get_unique_watched(user_data):
+    # tried to use set and .difference() method to solve,
+    # but finally found out need to return a list
+    friends_list = []
+    unique_list = []
+    friend_watched = user_data["friends"]
+    user_watched = user_data["watched"]
+
+    for val in friend_watched:
+        # read all the "watched" in friends watched list
+        for movie in val["watched"]:
+            # read every title/genre/rating in each watched
+            friends_list.append(movie["title"])
+            # put the name of the movie into the empty list
+    for title in user_watched:
+        # read every movie in user watched list
+        if title["title"] not in friends_list: 
+            unique_list.append(title)
+    return unique_list   
+
+
+def get_friends_unique_watched(user_data):
+    user_watched_list = []
+    friends_unique_movies = []
+    friend_watched = user_data["friends"]
+    user_watched = user_data["watched"]
+
+    for title in user_watched:
+        user_watched_list.append(title["title"])

```

Comment:
This is not necessary! In line 118, instead of checking the title, you can check `if title not in user_watched`.